LaTeX: add `tectonic`

barucden opened the pull request:
Adds another option, "tectonic", to the `platform` argument of Latex writer. The option makes Latex writer use the tectonic binary to compile the tex files. If this is the way to go, I will adjust the documentation accordingly. 

I am not sure where to add a test case for this. I would appreciate help.

Closes #1802 

(cc @fredrikekre)

inkydragon commented on Project.toml:
```diff
@@ -16,6 +16,7 @@ Markdown = "d6f4376e-aef5-505a-96c1-9c027394607a"
 REPL = "3fa0cd96-eef1-5676-8a61-b3b8758bbffb"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 Unicode = "4ec0a83e-493e-50e2-b9ac-8f72acf5a8f5"
+tectonic_jll = "d7dd28d6-a5e6-559c-9131-7eb760cdacc5"
```
1. Can we make this dep optional? Not everyone wants to build a PDF.

Of course, if the optional dependencies are not easy to implement, adding a [~5MB][5MB] dependency is not unacceptable

2. Add `tectonic_jll` or [Tectonic.jl](https://github.com/MichaelHatherly/Tectonic.jl)?

[5MB]: https://github.com/JuliaBinaryWrappers/tectonic_jll.jl/releases/tag/tectonic-v0.8.0%2B0

fredrikekre replied:
I was going to comment the same. I think you can do what I did for node in https://github.com/JuliaDocs/Documenter.jl/pull/1627

barucden replied:
What should the interface look like then? `LaTeX(platform="native", tectonic=tectonic_jll.tectonic())`? Or `LaTeX(platform="tectonic", tectonic=tectonic_jll.tectonic())`? Or something else?

fredrikekre replied:
The latter, and passing `tectonic` can be ignored if you have it in `PATH` already.

mortenpi commented:
A test case should be added into `test/examples/make.jl` and `test/examples/tests_latex.jl`, and it should probably not run by default (we don't run LaTeX builds by default with the tests), but as part of [this CI job](https://github.com/JuliaDocs/Documenter.jl/blob/cd57c41ab1ff8a8dacf5c21a7d8363b777a6ad9a/.github/workflows/CI.yml#L45-L68), with the Tectonic JLL added into the environment.

barucden commented:
The interface now accepts a path to `tectonic` or takes it from `PATH`. I added a test-case as per @mortenpi's suggestion, and adjusted the CI job accordingly. Lastly, I updated the documentation of LaTeX writer.

rikhuijzer commented on src/Writers/LaTeXWriter.jl:
```diff
@@ -180,6 +188,19 @@ function compile_tex(doc::Documents.Document, settings::LaTeX, texfile::String)
+            piperun(`$(tectonic) -X compile --keep-logs -Z shell-escape $texfile`)
+            return true
+        catch err
+            logs = cp(pwd(), mktempdir(); force=true)
```
```suggestion
            logs = cp(pwd(), mktempdir(); force=true)
```
Isn't that dir deleted when Julia exits? Can you also print the output for people who want to debug CI logs? Maybe also use the `--print` flag. That one has helped me a lot in some situations when debugging Books.jl.

barucden replied:
Yes, according to the docs, the temporary directory is deleted when Julia exits. Should we fix it also for the "native" (latexmk) case?

What output are you referring to? That of the `tectonic` command? If I understand `piperun` correctly, the output is visible if you pass `--verbose` to `ARGS`. Otherwise, it gets redirected to files.

rikhuijzer replied:
Yes, I meant for `tectonic`.

rikhuijzer replied:
> Should we fix it also for the "native" (latexmk) case?

I don't know. I only know things about Tectonic and not much about Documenter

fredrikekre replied:
#1697

rikhuijzer commented on test/examples/make.jl:
```diff
@@ -441,6 +441,26 @@ else
 
+examples_latex_simple_tectonic_doc = if "latex_simple_tectonic" in EXAMPLE_BUILDS
+    @info("Building mock package docs: LaTeXWriter/latex_simple_tectonic")
+    import tectonic_jll: tectonic
```
```suggestion
    using tectonic_jll: tectonic
```

barucden commented:
I missed a space in `ENV["DOCUMENTER_TEST_EXAMPLES"]`. The tests should be fine now.

fredrikekre commented on docs/src/man/other-formats.md:
```diff
@@ -33,6 +33,27 @@ The following is required to build the documentation:
+### Compiling using Tectonic 
+
+The documentation can be also built using the
+[Tectonic](https://tectonic-typesetting.github.io) typesetting system. It is required to have a `tectonic`
```
```suggestion
[Tectonic](https://tectonic-typesetting.github.io) LaTeX engine. It is required to have a `tectonic`
```

fredrikekre commented on docs/src/man/other-formats.md:
```diff
@@ -33,6 +33,27 @@ The following is required to build the documentation:
+
+The documentation can be also built using the
+[Tectonic](https://tectonic-typesetting.github.io) typesetting system. It is required to have a `tectonic`
+executable present in `PATH`, or to provide it using the `tectonic` keyword:
```
```suggestion
available in `PATH`, or to provide a path to the binary using the `tectonic` keyword:
```

fredrikekre commented on docs/src/man/other-formats.md:
```diff
@@ -33,6 +33,27 @@ The following is required to build the documentation:
+    format = Documenter.LaTeX(platform="tectonic"), 
+    ...)
+
+# The path to `tectonic` is provided by the JLL
```
```suggestion
# The path to `tectonic` is provided by the tectonic_jll package
```

fredrikekre commented on docs/src/man/other-formats.md:
```diff
@@ -33,6 +33,27 @@ The following is required to build the documentation:
+    ...)
+
+# The path to `tectonic` is provided by the JLL
+import tectonic_jll: tectonic
```
```suggestion
using tectonic_jll: tectonic
```

fredrikekre commented on test/examples/tests_latex.jl:
```diff
@@ -3,7 +3,8 @@ using Test
 # make.jl, and we need to set it to the relevant LaTeX builds.
 ENV["DOCUMENTER_TEST_EXAMPLES"] =
-    "latex latex_simple latex_texonly latex_cover_page latex_toc_style"
+    "latex latex_simple latex_texonly latex_cover_page latex_toc_style " *
+    "latex_simple_tectonic"
```
```suggestion
    "latex latex_simple latex_texonly latex_cover_page latex_toc_style latex_simple_tectonic"
```

fredrikekre commented:
Can you also add an entry in the changelog?
